stable_baselines3/common/pytree_dataclass.py

```diff
@@ -1,21 +1,243 @@
+from typing_extensions import dataclass_transform
 
-__all__ = ["tree_flatten", "PyTree"]
+from stable_baselines3.common.type_aliases import TensorIndex
+from stable_baselines3.common.utils import zip_strict
 
+__all__ = [
+    "PyTreeDataclass",
+    "MutablePyTreeDataclass",
+    "TensorTree",
+    "tree_empty",
+    "tree_flatten",
+    "tree_index",
+    "tree_map",
+]
+
+S = TypeVar("S")
 T = TypeVar("T")
+U = TypeVar("U")
 
 SB3_NAMESPACE = "stable-baselines3"
 
 
+# We need to inherit from `type(CustomTreeNode)` to prevent conflicts due to different-inheritance in metaclasses.
+# - For some reason just inheriting from `typing._ProtocolMeta` does not get rid of that error.
+# - Inheriting from `typing._GenericAlias` is impossible, as it's a `typing._Final` class.
+#
+# But in mypy, inheriting from a dynamic base class from `type` is not supported, so we disable type checking for this
+# line.
+class _PyTreeDataclassMeta(type(CustomTreeNode)):  # type: ignore[misc]
+    """Metaclass to register dataclasses as PyTrees.
+
+    Usage:
+      class MyDataclass(metaclass=_DataclassPyTreeMeta):
+        ...
+    """
+
+    currently_registering: ClassVar[Optional[type]] = None
+
+    def __new__(mcs, name, bases, namespace, slots=True, **kwargs):
+        cls = super().__new__(mcs, name, bases, namespace)
+
+        if dataclasses.is_dataclass(cls):
+            if mcs.currently_registering is not None:
+                assert mcs.currently_registering.__module__ == cls.__module__
+                assert mcs.currently_registering.__qualname__ == cls.__qualname__
+                mcs.currently_registering = None
+                return cls
+            else:
+                assert cls.__name__ in ["PyTreeDataclass", "MutablePyTreeDataclass"] or issubclass(
+                    cls, (PyTreeDataclass, MutablePyTreeDataclass)
+                )
+                mcs.currently_registering = cls
+        else:
+            mcs.currently_registering = cls
+
+        if name != "_PyTreeDataclassBase":
+            if name not in ["PyTreeDataclass", "MutablePyTreeDataclass"]:
+                frozen = issubclass(cls, PyTreeDataclass)
+                if frozen:
+                    assert not issubclass(cls, MutablePyTreeDataclass)
+                else:
+                    assert issubclass(cls, MutablePyTreeDataclass)
+            else:
+                frozen = kwargs.pop("frozen")
+
+            cls = dataclasses.dataclass(frozen=frozen, slots=slots, **kwargs)(cls)
+            assert issubclass(cls, CustomTreeNode)
+            ot.register_pytree_node_class(cls, namespace=SB3_NAMESPACE)
+        return cls
+
+
+class _PyTreeDataclassBase(CustomTreeNode[T], metaclass=_PyTreeDataclassMeta):
+    _names_cache: ClassVar[Optional[Tuple[str, ...]]] = None
+
+    # Mark this class as a dataclass, for type checking purposes.
+    __dataclass_fields__: ClassVar[Dict[str, dataclasses.Field[Any]]]
+
+    @classmethod
+    def _names(cls) -> Tuple[str, ...]:
```

Comment:
Seems like this could be simplified to:
```
        if cls._names_cache is None:
            cls._names_cache = tuple(f.name for f in dataclasses.fields(cls))
        return cls._names_cache
```